libcnb-test/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+}
+
+/// References a Cloud Native Buildpack
+#[derive(Eq, PartialEq, Debug)]
+pub enum BuildpackReference {
+    /// References the buildpack in the Rust Crate currently being tested
+    Crate,
+    /// References another buildpack by id, local directory or tarball
+    Other(String),
+}
+
+impl IntegrationTest {
+    /// Creates a new integration test.
+    ///
+    /// # Panics
+    /// - When the connection to Docker failed
+    /// - When the internal Tokio runtime could not be created
+    pub fn new(builder_name: impl Into<String>, app_dir: impl AsRef<Path>) -> Self {
+        IntegrationTest {
+            app_dir: PathBuf::from(app_dir.as_ref()),
+            target_triple: String::from("x86_64-unknown-linux-musl"),
+            builder_name: builder_name.into(),
+            buildpacks: vec![BuildpackReference::Crate],
+            docker: Docker::connect_with_local_defaults()
+                .expect("Could not connect to local Docker deamon"),
+            tokio_runtime: tokio::runtime::Runtime::new()
+                .expect("Could not create internal Tokio runtime"),
+        }
+    }
+
+    /// Sets the buildpacks order.
+    ///
+    /// Defaults to [`BuildpackReference::Crate`].
+    pub fn buildpacks(&mut self, buildpacks: impl Into<Vec<BuildpackReference>>) -> &mut Self {
+        self.buildpacks = buildpacks.into();
+        self
+    }
+
+    /// Sets the target triple.
+    ///
+    /// Defaults to `x86_64-unknown-linux-musl`.
+    pub fn target_triple(&mut self, target_triple: impl Into<String>) -> &mut Self {
+        self.target_triple = target_triple.into();
+        self
+    }
+
+    /// Starts a new integration test run.
+    ///
+    /// This function will copy the application to a temporary directory, cross-compiles this crate,
+    /// packages it as a buildpack and then invokes [pack](https://buildpacks.io/docs/tools/pack/)
+    /// to build a new Docker image with the buildpacks specified by this integration test instance.
+    ///
+    /// Since this function is supposed to only be used in integration tests, failures are not
+    /// signalled via [`Result`](Result) values. Instead, this function panics whenever an unexpected error
+    /// occurred to simplify testing code.
+    ///
+    /// # Panics
+    /// - When the app could not be copied
+    /// - When this crate could not be packed as a buildpack
+    /// - When the "pack" command unexpectedly fails
+    ///
+    /// # Example
+    /// ```no_run
+    /// use libcnb_test::IntegrationTest;
+    ///
+    /// IntegrationTest::new("heroku/buildpacks:20", "test-fixtures/app")
+    ///     .run(|context| {
+    ///         assert!(context.pack_stdout.contains("---> Ruby Buildpack"));
+    ///         assert!(context.pack_stdout.contains("---> Installing bundler"));
+    ///         assert!(context.pack_stdout.contains("---> Installing gems"));
+    ///     })
+    /// ```
+    pub fn run<F: FnOnce(IntegrationTestContext)>(&mut self, f: F) {
```

Comment:
```suggestion
    pub fn build<F: FnOnce(IntegrationTestContext)>(&mut self, f: F) {
```

Maybe this is a me thing, but when I read `run` I think it both builds and runs the container. I wonder if something like `build` that matches `pack build` would make sense?

Reply:
My reasoning was that you "run" the integration test. Happy to change the name though.

Reply:
I agree that `run` could mean a few different things. However I wonder if `build` is a bit too much of an overloaded word, given it could mean any of:
- build the buildpack only
- build the buildpack and run pack build
- run pack build only

In addition, this function runs the passed function in addition to compiling/packaging/pack-building, which seems outside the scope of "build".

Manual mentioned `run_test` on Slack, which I quite like, since it gives more context.

Reply:
I'll change it to `run_test` for now.

Reply:
Renamed in 10328d1